src/libstd/thread/mod.rs

```diff
@@ -525,6 +526,35 @@ pub fn park_timeout(dur: Duration) {
 }
 
 ////////////////////////////////////////////////////////////////////////////////
+// ThreadId
+////////////////////////////////////////////////////////////////////////////////
+
+/// A unique identifier for a running thread.
+///
+/// A `ThreadId` is an opaque object that has a unique value for each thread
+/// that creates one. `ThreadId`s do not correspond to a thread's system-
+/// designated identifier.
+#[unstable(feature = "thread_id", issue = "21507")]
+#[derive(Eq, PartialEq, Copy, Clone)]
+pub struct ThreadId(usize);
+
+impl ThreadId {
+    /// Returns an identifier unique to the current calling thread.
+    #[unstable(feature = "thread_id", issue = "21507")]
+    pub fn current() -> ThreadId {
+        static THREAD_ID_COUNT: AtomicUsize = ATOMIC_USIZE_INIT;
+        #[thread_local] static mut THREAD_ID: ThreadId = ThreadId(0);
+
+        unsafe {
+            if THREAD_ID.0 == 0 {
+                THREAD_ID.0 = 1 + THREAD_ID_COUNT.fetch_add(1, Ordering::SeqCst);
+            }
+            THREAD_ID
```

Comment:
Not that Java is something I look up to, but here's how Java implements unique thread IDs:

```java
private static synchronized long nextThreadID() {
    return ++threadSeqNumber;
}
```

My gut intuition (without doing any calculating) is that It would take thousands of years of running for a loop like that spawning threads to overflow 64 bits. Highly unlikely.